shared/src/main/java/com/couchbase/lite/Replicator.java

```diff
@@ -0,0 +1,617 @@
+    }
+
+    /**
+     * The replicator's configuration.
+     *
+     * @return
+     */
+    public ReplicatorConfiguration getConfig() {
+        return config.copy();
+    }
+
+    /**
+     * The replicator's current status: its activity level and progress. Observable.
+     *
+     * @return
+     */
+    public Status getStatus() {
+        return status.copy();
+    }
+
+    /**
+     * Set the given ReplicatorChangeListener to the this replicator.
+     *
+     * @param listener
+     */
+    public void addChangeListener(ReplicatorChangeListener listener) {
+        if (listener == null)
+            throw new IllegalArgumentException();
+        changeListeners.add(listener);
+    }
+
+    /**
+     * Remove the given ReplicatorChangeListener from the this replicator.
+     *
+     * @param listener
+     */
+    public void removeChangeListener(ReplicatorChangeListener listener) {
+        if (listener == null)
+            throw new IllegalArgumentException();
+        changeListeners.remove(listener);
+    }
+
+    @Override
+    public String toString() {
+        if (desc == null)
+            desc = description();
+        return desc;
+    }
+
+    //---------------------------------------------
+    // Implementation of NetworkReachabilityListener
+    //---------------------------------------------
+    @Override
+    public void networkReachable() {
+        if (c4repl == null) {
+            Log.e(TAG, "%s: Server may now be reachable; retrying...", this);
+            retryCount = 0;
+            retry();
+        }
+    }
+
+    @Override
+    public void networkUnreachable() {
+        Log.v(TAG, "%s: Server may NOT be reachable now.", this);
+    }
+
+    //---------------------------------------------
+    // Protected level access
+    //---------------------------------------------
+
+    // - (void) dealloc
+    @Override
+    protected void finalize() throws Throwable {
+        clearRepl();
+        if (reachabilityManager != null) {
+            reachabilityManager.removeNetworkReachabilityListener(this);
+            reachabilityManager = null;
+        }
+        super.finalize();
+    }
+
+    //---------------------------------------------
+    // Private member methods (in class only)
+    //---------------------------------------------
+
+    private void _start() {
+        // Source
+        com.couchbase.litecore.Database db = config.getDatabase().internal();
+
+        // Target:
+        String schema = null;
+        String host = null;
+        int port = 0;
+        String path = null;
+
+        URI remoteURI = config.getTarget().getUri();
+        String dbName = null;
+        com.couchbase.litecore.Database otherDB = null;
+        // replicate against remote endpoint
+        if (remoteURI != null) {
+            schema = remoteURI.getScheme();
+            host = remoteURI.getHost();
+            port = remoteURI.getPort();
```

Comment:
https://docs.oracle.com/javase/7/docs/api/java/net/URI.html#getPort()

I should confirm how it behaves.